### ISSUE TITLE
Debug task is run even though dependent task is skipped

### ISSUE THREAD

**Issue by agaffney** [ansible/ansible#13161]
I found a case in ansible 1.9.4 (also present in 1.9.2) where a debug task with `when: whatever|success` runs even though the task that registers `whatever` was skipped.

I was able to reproduce with this playbook:

```
- hosts: all
  user: cloud-user

  tasks:
    - name: command that always fails
      shell: /bin/false
      ignore_errors: yes
      register: false_command

    - name: dummy git task
      git: accept_hostkey=yes
      when: false_command|success
      register: dummy_git

    - name: dummy debug
      debug:
        msg: "This should never happen"
      when: dummy_git|success
```

which results in the following output:

```
$ ansible-playbook -i 10.60.3.31, ansible_problem.yml 

PLAY [all] ******************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [10.60.3.31]

TASK: [command that always fails] ********************************************* 
failed: [10.60.3.31] => {"changed": true, "cmd": "/bin/false", "delta": "0:00:00.002888", "end": "2015-11-13 11:26:25.460150", "rc": 1, "start": "2015-11-13 11:26:25.457262", "warnings": []}
...ignoring

TASK: [dummy git task] ******************************************************** 
skipping: [10.60.3.31]

TASK: [dummy debug] *********************************************************** 
ok: [10.60.3.31] => {
    "msg": "This should never happen"
}

PLAY RECAP ******************************************************************** 
10.60.3.31                 : ok=3    changed=1    unreachable=0    failed=0   
```

**Comment by agaffney** [comment]
I think I see the problem. The 'success' filter uses the 'failed' filter, which explicitly looks for 'rc' or 'failed' fields, and otherwise assumes success. The "empty" dict from a skipped resource has neither, so is considered successful.

**Comment by agaffney** [comment]
I also tried changing the `when` clause on the debug task to the following with the same result:

```
when: dummy_git is defined
```

The next attempt that checks for a specific field inside the registered variable actually seemed to work:

```
when: dummy_git.after is defined
```

**Comment by bcoca** [comment]
discussing in core if we should switch success to be 'not failed and not skipped'

**Comment by bcoca** [comment]
if you want to test the possible patch:

```
diff --git a/lib/ansible/plugins/test/core.py b/lib/ansible/plugins/test/core.py
index 1bd7892..06fa687 100644
--- a/lib/ansible/plugins/test/core.py
+++ b/lib/ansible/plugins/test/core.py
@@ -36,7 +36,7 @@ def failed(*a, **kw):

 def success(*a, **kw):
     ''' Test if task result yields success '''
-    return not failed(*a, **kw)
+    return not failed(*a, **kw) and not skipped(*a, **kw)

 def changed(*a, **kw):
     ''' Test if task result yields changed '''
```

**Comment by agaffney** [comment]
That patch results in the debug task being skipped (as expected) with `when: dummy_git|success`

**Comment by bcoca** [comment]
also backported fix to stable-1.9, jic there ever is a 1.9.5

**Comment by resmo** [comment]
also see PR #8907 and issue #8753
